srk/src/termination.ml

```diff
@@ -0,0 +1,429 @@
+      let ith_ele = Vec.coeff j row_i in
+      let term_v = mk_real srk ith_ele in
+      let prod = mk_mul srk [ List.nth lambdas j; term_v] in
+      let value = mk_add srk [ prod; !rlhs] in
+      rlhs := value
+    done;
+    r := List.concat [!r; [!rlhs]]
+  done;
+  !r
+
+let build_gt_zero srk lambdas =
+  List.fold_left
+    (fun f lambda ->
+      mk_and srk [f; mk_leq srk (mk_zero srk) lambda]
+    )
+    (mk_true srk)
+    lambdas
+
+let rec build_subtracted_lambdas srk lambdas1 lambdas2 =
+  match lambdas1, lambdas2 with
+  | h1 :: t1, h2 :: t2 -> 
+    mk_sub srk h1 h2 :: build_subtracted_lambdas srk t1 t2
+  | [], [] -> []
+  | _, _ -> assert false
+
+let get_constants_vec cs constraints =
+  BatList.fold_left
+    (fun (existing_rows, i) eq -> 
+        match eq with
+        | `LeqZero t -> 
+          begin
+            let row = Vec.coeff CoordinateSystem.const_id t in
+            let neg = QQ.negate row in
+            let m =  Vec.add_term neg i existing_rows in
+            (m, i+1)
+          end
+        | `LtZero t -> Log.error "there should not be strict inequalities"; assert false
+        | `EqZero t -> 
+          begin
+            let row = Vec.coeff CoordinateSystem.const_id t in
+            let neg = QQ.negate row in
+            let m =  Vec.add_term neg i existing_rows in
+            let row2 = Vec.coeff CoordinateSystem.const_id t in
+            let m =  Vec.add_term row2 (i+1) m in
+            (m, i+2)
+          end
+    )
+    (Vec.zero, 0)
+    constraints
+  
+let print_formula_to_stdout srk descrip f =
+  logf descrip;
+  let s2 = (Printf.sprintf "\n%s\n" (Formula.show srk f)) in
+  logf "%s" s2;
+  ()
+
+let print_ranking_func srk cs xp_list rf_terms_list rf_delta0 rf_delta interp =
+  (* let len = List.length xp_list in *)
+  let rf = List.map
+      (fun term -> Interpretation.evaluate_term interp term)
+      rf_terms_list
+  in
+  let d0 = List.map
+    (fun term -> Interpretation.evaluate_term interp term)
+    rf_delta0
+  in
+  let d = List.map
+    (fun term -> Interpretation.evaluate_term interp term)
+    rf_delta
+  in
+  logf "\n\n********* Ranking function ************\n";
+  Format.fprintf Format.std_formatter "f = ";
+  let n_cols = List.length rf in
+  for i = 0 to n_cols-1 do
+    let real_var = List.nth xp_list i in
+    (* let xp = List.nth xp_list i in *)
+    pp_symbol srk Format.std_formatter real_var;
+    Format.fprintf Format.std_formatter " * ";
+    let coe = List.nth rf i in
+    QQ.pp Format.std_formatter coe;
+    if i < n_cols-1 then
+    Format.fprintf Format.std_formatter " + ";
+  done;
+  
+  logf "\nIt has to decrese by this amount in each iter: %a" QQ.pp (QQ.sub QQ.zero (List.nth d 0)); 
+
+  (* let t = QQ.negate (List.nth d0 0) in *)
+  let t = QQ.sub (List.nth d 0) (List.nth d0 0) in 
+  logf "\nAnd its lower bound is: %a \n" QQ.pp t;
+  logf "\n\n***************************\n\n";
+  ()
+
+let prove_termination srk tto_transition_formula tmul loop =
```

Comment:
Parameterizing by tto_transition_formula is a bit weird, and tmul seems to be unused.
I think it would preferable to just take in a "stem" formula, a "loop" formula, and a list of symbol pairs.

Reply:
Also, the function should return an ranking function (option), not print it:
- There are lots of applications where we want to use termination proving as a sub-procedure
- Even if we only care about termination, when we run on termination benchmarks we want to decide whether the whole program terminates, which means we need to know whether prove_termination succeeds for *every* loop